### ISSUE TITLE
Challenges with install with an authenticated proxy

### ISSUE THREAD

**Issue by byjrack** [Homebrew/brew#15584]
### `brew doctor` output

```shell
Your system is ready to brew.
```

### Verification

- [X] My "`brew doctor` output" above says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` twice and am still able to reproduce my issue.
- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### `brew config` output

```shell
HOMEBREW_VERSION: 4.0.24
ORIGIN: https://github.com/Homebrew/brew
HEAD: 54c8876dc39047c04de15e7d212979ae8d98cf1c
Last commit: 4 days ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: ecd1fac6f6cbd13d825ef9fce65ca308c383d467
Core tap last commit: 43 minutes ago
Core tap branch: master
Core tap JSON: 23 Jun 12:06 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_REPOSITORY: /opt/homebrew/Homebrew
HOMEBREW_CASK_OPTS: ["--appdir=~/Applications"]
HOMEBREW_MAKE_JOBS: 10
http_proxy: http://:@proxy.example.com:8000
https_proxy: http://:@proxy.example.com:8000
no_proxy: 192.168.0.0/16,172.16.0.0/12,10.0.0.0/8,example.com
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_blizzard_avalanche
Clang: 14.0.3 build 1403
Git: 2.39.2 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.88.1 => /usr/bin/curl
macOS: 13.4-arm64
CLT: 14.3.1.0.1.1683849156
Xcode: N/A
Rosetta 2: false
```

### What were you trying to do (and why)?

Install formulas

### What happened (include all command output)?

My gut says something is broken down in the Negotiate support that I can't quite tease out.  Historically I would just be fine with either using the proxy url format http://:@proxy:port or a curlrc that included a --negotiate and --U: param, but that doesn't seem to be the case any longer.  

```
brew install python3 
==> Downloading https://formulae.brew.sh/api/formula.jws.json
curl: (22) The requested URL returned error: 403                               

Warning: formula.jws.json: update failed, falling back to cached version.
==> Downloading https://formulae.brew.sh/api/cask.jws.json
curl: (22) The requested URL returned error: 403                               

Warning: cask.jws.json: update failed, falling back to cached version.
```

I can correlate the api hits to proxy events
```
2023-06-23 11:49:51  -  DENIED CONNECT tcp://formulae.brew.sh:443/ 403 - "Technology/Internet" - configuration_error TCP_DENIED "Homebrew/4.0.24 (Macintosh; arm64 Mac OS X 13.4) curl/7.88.1" 22 206 237 tcp formulae.brew.sh 443 / - - - 185.199.109.153 443 - 8000 BASIC 0 HTTP/1.1 
```

```
==> Fetching dependencies for python@3.11: mpdecimal, readline, sqlite and xz
==> Fetching mpdecimal
==> Downloading https://ghcr.io/v2/homebrew/core/mpdecimal/manifests/2.5.1
Already downloaded: .../Library/Caches/Homebrew/downloads/f367c2ee08c56b88be0662703a8e4275f8657608a268c8c44e845154b0cea543--mpdecimal-2.5.1.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/mpdecimal/manifests/2.5.1
Error: python@3.11: Failed to download resource "mpdecimal_bottle_manifest"
Failure while executing; `/usr/bin/env /opt/homebrew/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.0.24\ \(Macintosh\;\ arm64\ Mac\ OS\ X\ 13.4\)\ curl/7.88.1 --header Accept-Language:\ en --retry 3 --header Accept:\ application/vnd.oci.image.index.v1\+json --header Authorization:\ Basic\ ****** --fail --location --silent --head --request GET https://ghcr.io/v2/homebrew/core/mpdecimal/manifests/2.5.1` exited with 22. Here's the output:
curl: (22) The requested URL returned error: 403
```

And for the above
```
2023-06-23 12:06:33  -  DENIED CONNECT tcp://ghcr.io:443/ 403 - "Technology/Internet" - configuration_error TCP_DENIED "Homebrew/4.0.24 (Macintosh; arm64 Mac OS X 13.4) curl/7.88.1" 59 206 219 tcp ghcr.io 443 / - - - 140.82.114.33 443 - 8000 BASIC 0 HTTP/1.1 
```

Using HOMEBREW_CURL_VERBOSE I can see a blank username and password is sent to the Proxy via Basic which won't work.  

```
* Establish HTTP proxy tunnel to formulae.brew.sh:443
* Proxy auth using Basic with user ''
> CONNECT formulae.brew.sh:443 HTTP/1.1
> Host: formulae.brew.sh:443
> Proxy-Authorization: Basic Og==
```

Using HOMEBREW_CURLRC I can get brew to support Negotiate so seems like a solid workaround when curlrc is configured correctly.  Any other recommendations on authenticated proxy support that support Negotiate?  Especially with the move to the REST API and away from git I am having to troubleshoot this more often w our users.

### What did you expect to happen?

Things download happily

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
every formula is having the same result, but brew install python3 as an example
```

**Comment by halamix2** [comment]
I got the same problem running `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"` helped, I think this might be caused by #15582

**Comment by byjrack** [comment]
Cool I will poke at that a bit.  I was figuring it was some control logic that saw the ":@" prefix and that fell into a standard basic auth flow.  Just helps me reduce the amount of prereqs folks need to have on their machines.

**Comment by MikeMcQuaid** [comment]
> I think this might be caused by #15582

on the off chance: CC @Bo98

**Comment by carlocab** [comment]
> I think this might be caused by #15582

I'm sceptical, TBH. This smells more like #15579, but I'll admit to not having taken a careful look.

**Comment by Bo98** [comment]
Shouldn't be either because:

> ```
> HOMEBREW_VERSION: 4.0.24
> ORIGIN: https://github.com/Homebrew/brew
> HEAD: 54c8876dc39047c04de15e7d212979ae8d98cf1c
> Last commit: 4 days ago
> ```

**Comment by byjrack** [comment]
Still no luck firguring out where the auth process changed for https://github.com/Homebrew/brew/commits/master/Library/Homebrew/utils/curl.rb.

Workaround for using HOMEBREW_CURLRC works with --negotiate, but trying to figure out how to handle the "-U :" situation better in the code.  I know the manpage says to encode basic in the proxy url, but feels like we can handle negotiate situations as well.

**Comment by MikeMcQuaid** [comment]
Closing as there's a workaround here. Will review PR(s) to better improve the situation.

**Comment by byjrack** [comment]
@MikeMcQuaid could use a pointer to where the meta object is created to think through what a PR would look like.  I believe using the "empty user" syntax in the proxy URL is triggering the [--user logic in the curlddownloadstrategy](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/download_strategy.rb#L561) which is why an empty basic auth header is added, but I can't trace where that hash is actually created.  Could be a wild goose chase, but may help me with the control logic.

**Comment by MikeMcQuaid** [comment]
@byjrack sorry, I'm not sure what is meant by "meta object" here?

**Comment by byjrack** [comment]
Well the hash is passed into the constructor, but what I trying to walk back is where is the hash defined so I can figure out if the param that ends up in "--user" may come from some parsing or the proxy environment var content or something else completely.

**Comment by byjrack** [comment]
Got it super helpful.  So the formula author can set some required curl args based on their specific needs and that function takes care of adding them to the command used in the downloadstrategy.  

Doesn't answer how I am ending up with a basic auth header in the curl requests, but closes off that path.  My gut is telling me its how curl is responding to the command provided by brew combined w the 407 coming back.  I just need to confirm that as I don't see any parsing of the proxy envvar content in brew that would tack on that header.

**Comment by carlocab** [comment]
From the error message you post above, the failing `curl` invocation (reformatted for readability) was
```
curl \
  --disable \
  --cookie /dev/null \
  --globoff \
  --show-error \
  --user-agent Homebrew/4.0.24\ \(Macintosh\;\ arm64\ Mac\ OS\ X\ 13.4\)\ curl/7.88.1 \
  --header Accept-Language:\ en \
  --retry 3 \
  --header Accept:\ application/vnd.oci.image.index.v1\+json \
  --header Authorization:\ Basic\ ****** \
  --fail \
  --location \
  --silent \
  --head \
  --request GET \
  https://ghcr.io/v2/homebrew/core/mpdecimal/manifests/2.5.1
```
I don't see a `--user` flag, so I suspect `:user` is the wrong property to look for. There is an `Authorization` header, though.

**Comment by MikeMcQuaid** [comment]
> Got it super helpful. So the formula author can set some required curl args based on their specific needs and that function takes care of adding them to the command used in the downloadstrategy.

Yes.



> I don't see a `--user` flag, so I suspect `:user` is the wrong property to look for. There is an `Authorization` header, though.

Agreed. The best bet, I think, is to reproduce Homebrew's `curl` call (`--debug --verbose` may be useful for you here) in the terminal and then play around with arguments until you get the behaviour you want/need.

**Comment by byjrack** [comment]
I think it's all coming together.  

So we also use Artifactory as an internal package proxy, but when i was getting the 403s from 4.x I pulled out the HOMEBREW_ARTIFACT_DOMAIN envvar to get back to "out of box".  I bet in some of the tests I didn't pull the BASIC_AUTH_TOKEN which was getting added to the curl call.  It wasn't in my brew config output so I removed it at one point, but given I can't find any other "Authorization: Basic" use in brew that has to be the reason its there for some of the other output captures.  

Now the Proxy should just care about Proxy-Authorization headers, but I bet it tries to use Authorization absent any of the others for it's auth requirement which fails since the former contains a key for a completely different system.

Joys of system interrelations!
